### DISCUSSION TITLE
Best way to update (in real time) items in a list

### DISCUSSION THREAD

**Discussion by tomasweigenast** [rrousselGit/riverpod#2876] [Q&A]
I'm creating a new app where a page renders a `ListView<ProductCard>`. The list gets the elements from a Riverpod's `Notifier`. Also, the same notifier (when created) registers a new listener to listen to product changes. `ProductCard` renders a `Product` class. 

The thing is, when the `Notifier` registers a change, for example, product `123` changed its price, should I:
1. Re-render the entire list changing the product that changed?
```dart
final list = state;
list.insertAt(list.indexOf(product), product);
state = list;
```

2. Use another package, like Provider, to create a per-ProductCard `ChangeNotifier` subclass and in there register a listener to listen for changes in the specific product, like:
```dart
class ProductNotifier extends ChangeNotifier {
    ProductNotifier(this.product) {
         listen(this.product.id, (update) => {});
    }

}
```
The `ListView<ProductCard>` should become `ListView<ChangeNotifierProvider<ProductCard>>`, then `ProductCard` should build a `Consumer`.

3. Maybe use `flutter_hooks` and register the listener inside a `useEffect` in the `ProductCard`.

I think with options 2 and 3, `List` in the main `Notifier` (which provides data to the `ListView`), will become stale.

What do you think guys? Is there any other alternative?

PD: Is good to mention that `ProductCard` can be rendered alone without a `ListView` on other pages in the same app.

**Comment by AhmedLSayed9** [comment]
You can scope the product id to have const widgets for the items inside your list view, and in the child widget use `.select` to filter rebuilds:
```dart
final produceId = ref.watch(scopedProductId);
final product = ref.watch(productListProvider.select((value) => value.firstWhere((e) => e.id == productId)));
```

**Reply by tomasweigenast** [reply]
So you say storing the id of the products in the Notifier instead of the Product itself? So, state becomes List<String> for example?

**Reply by AhmedLSayed9** [reply]
I'm not sure i follow up but you don't need another notifier.

**Reply by AhmedLSayed9** [reply]
Check [marvel example](https://github.com/rrousselGit/riverpod/blob/29f2f2eacf8f2e390b2522bc74fb12ca4a3495ee/examples/marvel/lib/src/screens/home.dart#L135) for how to scope the index/id and use it in the item/product widget.

For accessing the product and only rebuild when that product changes, you can either use `.select` or use another provider like the [example](https://github.com/rrousselGit/riverpod/blob/29f2f2eacf8f2e390b2522bc74fb12ca4a3495ee/examples/marvel/lib/src/screens/home.dart#L168).

**Comment by catalunha** [comment]
@AhmedLSayed9 @lucavenir 
Congratulations on the new and detailed text of docs-v2.
But I think a detailed and complete example of this should be there.
The use of ProviderScope for these cases was seen by me in an example but I still couldn't reproduce this fantastic resource of Riverpod with code quality.
And I believe he is of great community interest.
Taking advantage of the colleague's courage to post the help, could you help us with a detail of theory and code about this context of using ProviderScope in lists to update only one item and not the whole list?

**Reply by AhmedLSayed9** [reply]
There's a WIP topic for scoping providers at docs v2. Currently, You can check [marvel example](https://github.com/rrousselGit/riverpod/blob/29f2f2eacf8f2e390b2522bc74fb12ca4a3495ee/examples/marvel/lib/src/screens/home.dart#L135).

Scoping is a bit advanced and should be avoided if possible anyway. It achieves performance optimization by relying on const widgets but if the rebuild of your item isn't expensive, then it should be avoided imo.

**Reply by catalunha** [reply]
@AhmedLSayed9 
This information from you is very important. Because I thought that using ProviderScope would be the best way in these cases of dealing with lists.
So let's go to the summary of the list with lists:
1) User ProviderScope is not recommended
2) Better update the whole list.

**Reply by AhmedLSayed9** [reply]
I'm not saying it's not recommended nor it's better to update the whole list.
I'm saying it's a bit advanced and if you're not gaining noteworthy optimization, then it should be avoided. It's a case-by-case basis.

**Reply by lucavenir** [reply]
Scoping is an advanced topic, and you likely don't need scoping to solve your needs. I raised issues and discussions with Remi and basically it all came down to: scoping is there, you _can_ use it, but if anything I'd suggest to avoid it for now.

There're major implications when scoping providers. Stuff might not go as you expect. Also it introduces a lot of complexity.

FYI There's a proposal to introduce a "default family parameter", which combined with scoping lints would make scoping viable. But it's not there yet. The new documentation will _for sure_ specify how `.family` is all you need to deal with parameters.

Also: dealing with list, instead, usually don't involve families, nor scopes. I can't see why you'd need anything like that as of now, but maybe I didn't understand your problem.

Finally, as Ahmed said, no one can be sure what is best for _your_ use case. But as a rule of thumb, I'd update your `state` (assuming you've got a `Notifier` that exposes this `List<ProductCard>`) like so:

```dart
class MyListNotifier extends Notifier<List<MyEntity>> {
  @override
  List<MyEntity> build() {
    return [];
  }

  void updatePrice(int id, int price) {
    state = [
      ...state.map(
        (element) => element.id == id ? element.copyWith(price: price) : element,
      ),
    ];
  }
}

// example entity
@freezed
class MyEntity with _$MyEntity {
  const factory MyEntity({
    required int id,
    required int price,
  }) = _MyEntity;
}
```

Again *this is just an example we cannot be sure what you actually need* but at least I want to give a broader explanation of my POV

**Reply by tomasweigenast** [reply]
Your example is quite similar to what I'm trying to do actually. But the thing is that doing:
```dart
state = [
      ...state.map(
        (element) => element.id == id ? element.copyWith(price: price) : element,
      ),
    ];
```
for every update and if the list has many elements, it will have poor performance

**Reply by lucavenir** [reply]
> poor performance

Finding a better implementation is an exercise left to the reader 😄 

Jokes aside, if you're using immutable data, this functional approach is _very_ fast, as Dart recycles every other object reference that won't get mutated (aka any element s.t. `element.id != id`). Obviously, the comparison (`element.id == id`) is the only "expensive" operation here.

I doubt that such comparison will ever be a bottleneck for your application, but if it ever will be (assuming very large datasets), you can always pass an `index` (instead of an `id`) to the mutation. You'd obtain:
```dart
// immutability approach
void updatePrice(int index, int price) {
  state = [
    ...state.sublist(0, index),
    state[index].copyWith(price: price),
    ...state.sublist(index + 1, state.length),
  ];
}
```

If you're not into immutability (spoiler: a full article onto "why immutability?" is incoming), here's the full mutable alternative implementation:
```dart
class MyListNotifier extends Notifier<List<MyEntity>> {
  @override
  List<MyEntity> build() {
    return [];
  }

  void updatePrice(int index, int price) {
    state[index].price = price;
    ref.notifyListeners();
  }
}

// example *mutable* entity
@unfreezed
class MyEntity with _$MyEntity {
  const factory MyEntity({
    required int id,
    required int price,
  }) = _MyEntity;
}
```

**Reply by lucavenir** [reply]
Finally, consider using this external package to actually handle immutable lists and other collections: https://pub.dev/packages/fast_immutable_collections

(or any other good package such as this one)

**Reply by AhmedLSayed9** [reply]
@lucavenir
Great explanation! but I think he meant poor performance for the consumer widget, as all the list items will rebuild when only single item's price is updated (It's not a big deal if these widgets are not expensive).

For that, I think scoping and caching each item separately (using .select or family provider) to be able to rely on const widgets is the way to go.

**Reply by lucavenir** [reply]
I guess that's where immutability is also useful! I think that while the widget tree might execute `build` again, the element tree won't destroy its elements just because a sibling element mutated. I never experimented with this tho

**Comment by catalunha** [comment]
@AhmedLSayed9 
> I'm not saying it's not recommended nor it's better to update the whole list.

Forgiveness. I understand your speech. Mine is that it wasn't quite right.

@lucavenir 
Your explanations are always loaded with a lot of rich information. Thank you very much.
I know it's not the general case. But it covers a lot of common uses.
His example was also very important.
Thanks for spending time on it.
I suggest putting all your text from this issue in the documentation.
**Looking forward to every detail of the new documentation.**

Thank you very much. Guys

**Comment by catalunha** [comment]
> Jokes aside, if you're using immutable data, this functional approach is very fast, as Dart recycles every other object reference that won't get mutated (aka any element s.t. element.id != id). Obviously, the comparison (element.id == id) is the only "expensive" operation here.

Knowledge of this Dart's internal mechanism is very important. Congratulations.

**Comment by tomasweigenast** [comment]
Thank you everyone
